datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,412 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Distribution partitioner class that works with Hugging Face Datasets."""
+
+
+from collections import Counter
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArray, NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class DistributionPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on a distribution.
+
+    Inspired from implementations of Li et al. Federated Optimization in
+    Heterogeneous Networks (2020) https://arxiv.org/abs/1812.06127.
+
+    Given a user-specified distribution, the algorithm splits the dataset for each
+    unique label per partition where each label is assigned to the partitions in a
+    deterministic pathological manner. That is, given a distribution array of shape,
+                           `num_unique_labels_per_partition` x `num_partitions`
+    ( `num_unique_labels`, ---------------------------------------------------- ),
+                                          `num_unique_labels`
+    the label_id at the i'th row is assigned to the partition_id based on the formula:
+        partition_id = alpha + beta
+    where,
+        alpha* = (i - num_unique_labels_per_partition + 1) \
+                 + (j % num_unique_labels_per_partition)
+        alpha = alpha* + (alpha* > 0 ? 0 : num_unique_labels)
+        beta = num_unique_labels * (j // num_unique_labels_per_partition)
```

Comment:
> I checked for [1,1] in the described setup below, and it would assigned to the pid=3 (alpha=3; beta=0) but it is to the pid=1 (pid=1 aligned with my interpretation of the code)

Good catch, there's a typo there and should be `(alpha* >= 0 ? ...)`.

The `partition_id` array must first be generated for the i'th row for all `j` and then sorted in ascending order. 

I added the following sentence to be clear:
```
        partition_id = <alpha + beta>
    where,
        <.> denotes the reindexed sequence of partition_ids in monotone increasing
            order for all j's
```

Reply:
I've also added the alternative explanation.

Reply:
Sorry, I don't quite get it. I thought of alpha and beta as scalars.

Reply:
I'm not quite sure what got changed based on that
```
good catch, there's a typo there and should be (alpha* >= 0 ? ...).
```
Could you explain?